### PR TITLE
ENH: shortcut for positive definite matrices in SciPy.linalg.logm

### LINKED ISSUE #12464: linalg.logm slower than naive computation using svd
It seems the function linalg.logm is significantly slower than a naive way to compute the matrix logarithm using SVD. 

If we compute the matrix logarithm of a random positive definite matrix, I see a difference in performance of roughly factor 10 independent of matrix size. 

```
mat = np.random.normal(size=(1000,1000))
mat = mat@mat.T

U,S,V = np.linalg.svd(mat)
U@np.diag(np.log(S))@V
# takes 302ms on my machine

scipy.linalg.logm(mat)
# takes 2.79s on my machine
```

### PR DESCRIPTION
**Pull request by RikVoorhaar**
#### Reference issue
Closes #12464 

#### What does this implement/fix?
When computing the matrix logarithm of a positive definite matrix, we can take a shortcut. If A is PD and A = ZTZ* is the Schur decomposition, then T is diagonal (in general it is upper triangular). The matrix logarithm of T is then simply np.diag(np.log(T.diagonal())), and we don't need the machinery for computing the matrix log of an upper triangular matrix at all. 

I have therefore modified the method `linalg._matfuncs_inv_ssq.logm_triu` to check if the triangular matrix is actually a diagonal matrix, and if so use this simplified version of computing the log. This check has low overhead.

For depending on the size of the matrix, this can cause a speedup of up to a factor 10.

### CODE REVIEW

File: scipy/linalg/_matfuncs_inv_ssq.py [hunk 1]
```diff
@@ -760,13 +760,20 @@ def _logm_triu(T):
...
     keep_it_real = np.isrealobj(T) and np.min(T_diag) >= 0
     if keep_it_real:
         T0 = T
     else:
         T0 = T.astype(complex)
 
+    # Check if T0 is diagonal, if so return logarithm of the diagonal
+    # Reshape it so that diagonal becomes first columnn
+    T0_diagonal_reshape = T0.reshape(-1)[:-1].reshape(n-1,n+1)
+    is_diagonal = np.allclose(T0_diagonal_reshape[:,1:], 0)
```
Comment (pv) [inline scipy/linalg/_matfuncs_inv_ssq.py:773]: allclose has atol just 1e-8. The off-diagonal elements probably should be compared to the diagonal ones (which of them?) to determine whether the matrix is purely diagonal or not.

Comment (RikVoorhaar) [reply scipy/linalg/_matfuncs_inv_ssq.py:773]: Taking a peek at Higham section 11.2 about the conditioning of the matrix logarithm, we see that for positive definite matrices the conditioning number of log(A) is k(A)/||log(A)|| (and for non-PD this is a lower bound). Therefore if the off diagonal elements are small compared to the smallest diagonal element, it should be stable. (but this is not really a formal proof either)

Perhaps we can check whether the ratio of smallest diagonal element to biggest off-diagonal is bounded by e.g. 1e-8.

I can try to get a formal proof going

Comment (RikVoorhaar) [reply scipy/linalg/_matfuncs_inv_ssq.py:773]: I did some numerical experiments. We take a diagonal D matrix of norm 1, with entries following a linear scale (to create an arbitrarily high condition number). We perturb it with a uniform random upper triangular matrix P. 
We then record the relative error between logm(D) and logm(D+P) as function of the ratio between the minimal entry of D and maximal entry of P. The logarithms of the relative errors and this min/max ratio seem to follow a linear relationship.
In practice if this ratio is ~1e-8, then the relative error is also around 1e-8 (and in fact lower if the condition number is very high)

Therefore it seems that this ratio of smallest diagonal to biggest off-diagonal is a good indicator. 

Then we just have to fix a good value for this ratio. The goal is that the method is faster for PD matrices, so let's take the Schur decomposition of some PD matrices and record what this ratio is for the upper triangular matrix of the decomposition. It seems that the log of this ratio depends linearly on the log of the condition number of the PD matrix. 
At a condition number of ~1e7 the ratio is around 1e-7 to 1e-9. 

Even at very high condition number (>1e10), the we still see that approximation using only the diagonal is good; it both corresponds well to scipy's matrix logarithm, and if we exponentiate it we recover the matrix to a relative error of <1e-14. 

All in all, it seems that a ratio of 1e-8 is a reasonable tradeoff.

**Comment by RikVoorhaar** [comment]
Some build checks are failing, but from the logs I don't see any plausible relation to the code I actually changed, so I have no clue how to debug this.

**Comment by pv** [comment]
I think the 1e-8 is too large error, significantly closer to machine epsilon would be better (error preferably << 1e-12) --- in speed vs. precision tradeoffs I would choose precision, especially in cases where tolerance is not user-specified.

Here (with gh-12474), we are also talking about only ~2x speedup for matrices of sizes larger than n=200, diminishing somewhat for larger n, and for smaller n growing to ~5x at n=60 and to ~10x at n=20. OTOH, a structure assumption flag switching to eigh gives ~4-20x speedup in this, being e.g. 8x at n=200.

In schur with normal matrix input, numerical experiments seem to indicate deviation from exactly diagonal is probably of order ~ eps*max(abs(diag(T))), so whether the check correctly detects the structure depends on `|lambda_max|/|lambda_min| < eps/threshold`.

**Comment by ilayn** [comment]
This test makes me a bit uncomfortable since Schur decomposition has stability issues when eigenvalues are clustered close to zero. The typical way to check positive definiteness is via the success of the Cholesky factorization which is wasteful here since Cholesky won't be utilized afterwards. But at least for condition number, `?pocon` from LAPACK can be used instead of an overflow/underflow risk. 

However, a hard-coded number would definitely cause problems (as it does everywhere else) and I'm not sure if a user quickly would understand how to use this tolerance at all.

**Comment by RikVoorhaar** [comment]
well, in that case just adding a flag in logm is probably the most elegant solution.

**Comment by pv** [comment]
It might be better to add a `logm(a, assume_a='pos/her/nor')` flag that allows the user to tell if the input is PD/hermitian/normal, instead of `tol=`, which leaves the issue to the user and we can do things faster than schur. I'm sorry I didn't think about the tolerances when I commented on this change in #12464

**Comment by RikVoorhaar** [comment]
I changed it to use assumption flags instead. 

I still need to add tests for these new flags.

**Pull request closed without merging**